docs/docs/concepts/Sync_Async.md

```diff
@@ -0,0 +1,204 @@
+            var keys = res[1];
+            keys.forEach((key) => {
+                if (client.call('hget', key, 'name') == expected_name) {
+                    count += 1;
+                }
+            });
+        });
+    } while(cursor != '0');
+
+    // set count to the cache wil 5 seconds expiration
+    async_client.block((client)=>{
+        client.call('set', expected_name + '_count', count);
+        client.call('expire', expected_name + '_count', 5);
+    });
+
+    return count;
+});
+```
+
+The above code works as expected, it first check the cache, if cache exists it returns it, otherwise it is perform the calculation and update the cache. But the above example is not optimal, the callback is a Coroutine which means that it will always be calculated on a background thread. Moving to a background thread by itself is costly, the best approach would have been to check the cache synchronously and only if its not there, move to the background. Triggers and Functions allows to start synchronously and move asynchronously using `executeAsync` function. The new code:
+
+```js
+#!js api_version=1.0 name=lib
+
+redis.registerAsyncFunction('test', function(client, expected_name){
+    // check the cache first
+    var cached_value = client.call('get', expected_name + '_count');
+    if (cached_value != null) {
+        return cached_value;
+    }
+
+    // cache is not set, move to background
+    return client.executeAsync(async function(async_client) {
+        var count = 0;
+        var cursor = '0';
+        do{
+            async_client.block((client)=>{
+                var res = client.call('scan', cursor);
+                cursor = res[0];
+                var keys = res[1];
+                keys.forEach((key) => {
+                    if (client.call('hget', key, 'name') == expected_name) {
+                        count += 1;
+                    }
+                });
+            });
+        } while(cursor != '0');
+
+        // set count to the cache wil 5 seconds expiration
+        async_client.block((client)=>{
+            client.call('set', expected_name + '_count', count);
+            client.call('expire', expected_name + '_count', 5);
+        });
+
+        return count;
+    });
+});
+```
+
+`executeAsync` will return a `Promise` object, we return this Promise object as the function return value. When Triggers and Functions sees that the function returned a Promise, it waits for the promise to be resolved and return its result to the client. The above implementation will be much faster in case of cache hit.
+
+**Notice** that even though we registered a sync function (not a Coroutine) we still used `registerAsyncFunction`. This is because our function has the potential of blocking the client and take the execution to the background. If we would have used `registerFunction` Triggers and Functions would not have allow us to blocked the client and would have ignore the returned promise object.
```

Comment:
```suggestion
**Notice** that even though we registered a synchronous function (not a Coroutine) we still used `registerAsyncFunction`. This is because our function has the potential of blocking the client, taking the execution to the background. If we had used `registerFunction`, Triggers and Functions would not have allowed the function to block the client and it would have ignored the returned promise object.
```